server/rooms.ts

```diff
@@ -1253,6 +1253,72 @@ export class GlobalRoomState {
 		} catch (e) {}
 		this.lastBattle = Number(lastBattle) || 0;
 		this.lastWrittenBattle = this.lastBattle;
+		void this.loadBattles();
+	}
+
+	async saveBattles() {
+		let count = 0;
+		if (!Config.usepostgres) return 0;
+		const logDatabase = new PostgresDatabase();
+		for (const room of Rooms.rooms.values()) {
+			if (!room.battle || room.battle.ended) continue;
+			room.battle.frozen = true;
+			const log = await room.battle.getLog();
+			const players: ID[] = room.battle.options.players || [];
+			if (!players.length) {
+				for (const num of ['p1', 'p2', 'p3', 'p4'] as const) {
+					if (room.battle[num]?.id) {
+						players.push(room.battle[num].id);
+					}
+				}
+			}
+			if (!players.length || !log?.length) continue; // ???
+			await logDatabase.query(
+				`INSERT INTO stored_battles (roomid, input_log, players, title) VALUES ($1, $2, $3, $4)` +
+				` ON CONFLICT (roomid) DO UPDATE SET input_log = $2, players = $3, title = $4`,
```

Comment:
use `input_log = EXCLUDED.input_log` etc instead of rebinding the values, and maybe use SQL template strings instead of providing the array afterwards?